js/src/lib/Details/index.js

```diff
@@ -102,54 +103,152 @@ class Details extends Component {
-    ) {
-      if (this.state.changelogFilename) {
+  // Get repository details, like stars, changelog, commit activity and so on
+  getRepositoryDetails({ user, project, host, branch, path }) {
+    const { readme, changelogFilename } = this.state;
+    const hasReadme =
+      readme && readme.length > 0 && readme !== readmeErrorMessage;
+
+    if (host === 'github.com') {
+      if (changelogFilename) {
         get({
-          url: this.state.changelogFilename,
+          url: changelogFilename,
           type: 'text',
         }).then(res => this.setState({ changelog: res }));
       }
 
-      if (
-        typeof this.state.readme === 'undefined' ||
-        this.state.readme.length === 0 ||
-        this.state.readme === readmeErrorMessage
-      ) {
+      if (!hasReadme) {
+        this.setState({ readmeLoading: true });
         get({
           url: prefixURL('README.md', {
             base: 'https://raw.githubusercontent.com',
-            user: this.state.githubRepo.user,
-            project: this.state.githubRepo.project,
-            head: this.state.githubRepo.head
-              ? this.state.githubRepo.head
-              : 'master',
-            path: this.state.githubRepo.path.replace(/\/tree\//, ''),
+            user,
+            project,
+            head: branch,
+            path: path.replace(/\/tree\//, ''),
           }),
           type: 'text',
-        }).then(res => this.setState({ readme: res }));
+        })
+          .then(res => this.setState({ readme: res }))
+          .catch(() => this.setState({ readmeLoading: false }));
       }
 
       this.getGithub({
-        url: `repos/${this.state.githubRepo.user}/${
-          this.state.githubRepo.project
-        }/stats/commit_activity`,
+        url: `repos/${user}/${project}/stats/commit_activity`,
         state: 'activity',
       });
 
       this.getGithub({
-        url: `repos/${this.state.githubRepo.user}/${
-          this.state.githubRepo.project
-        }`,
+        url: `repos/${user}/${project}`,
         state: 'github',
       });
+    } else if (host === 'gitlab.com') {
+      const getGitlabFile = ({ user, project, branch, filePath }) => {
+        // We need to use the Gitlab API because the raw url does not support cors
+        // https://gitlab.com/gitlab-org/gitlab-ce/issues/25736
+        // So we need to 'translate' raw urls to api urls.
+        // E.g (https://gitlab.com/janslow/gitlab-fetch/raw/master/CHANGELOG.md) -> (https://gitlab.com/api/v4/projects/janslow%2Fgitlab-fetch/repository/files/CHANGELOG.md?ref=master)
+        // Once gitlab adds support, we can get rid of this workaround.
+        const apiUrl = `https://gitlab.com/api/v4/projects/${user}%2F${project}/repository/files/${encodeURIComponent(
+          filePath
+        )}?ref=${branch}`;
+        return get({
+          url: apiUrl,
+          type: 'json',
+        }).then(res => res.encoding === 'base64' && atob(res.content));
```

Comment:
this seems like a bit much code in this one file, what about extracting it into a js file?

Reply:
You mean the `getRepositoryDetails` function?

Reply:
yes, I wonder if it can be extracted. Maybe relying too much on `this` though, instead it can be passed `repository` maybe?

Reply:
What should we do about the `setState`'s ? 
Should we use a reference to the Details component or wait until we fetch all information and return a new state (Promise)?